proxy/__init__.py

```diff
@@ -0,0 +1,193 @@
+"""
+The proxy module provides the means to proxy images from other registry instances.
+Registries following the distribution spec are supported.
+"""
+from __future__ import annotations
+from typing import Any
+import re
+import json
+
+import requests
+
+from app import model_cache
+from data.cache import cache_key
+from data.database import ProxyCacheConfig
+from digest.digest_tools import sha256_digest
+
+
+WWW_AUTHENTICATE_REGEX = re.compile(r'(\w+)[=] ?"?([^",]+)"?')
+TOKEN_VALIDITY_LIFETIME_S = 60 * 60  # 1 hour, in seconds - Quay's default
+TOKEN_RENEWAL_THRESHOLD = 10  # interval (in seconds) when to renew auth token
+
+
+REGISTRY_URLS = {"docker.io": "registry-1.docker.io"}
+
+
+class UpstreamRegistryError(Exception):
+    def __init__(self, detail):
+        msg = f"Error requesting upstream registry: {detail}"
+        super().__init__(msg)
+
+
+def parse_www_auth(value: str) -> dict[str, str]:
+    """
+    Parses WWW-Authenticate parameters and returns a dict of key=val.
+    See https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/WWW-Authenticate
+    for details.
+    This parser is *not* fully compliant with RFC 7235, notably it does not support
+    multiple challenges.
+    """
+    scheme = value.split(" ", 1)[0]
+    matches = WWW_AUTHENTICATE_REGEX.findall(value)
+    parsed = dict(matches)
+    if scheme:
+        parsed["scheme"] = scheme
+    return parsed
+
+
+class Proxy:
+    def __init__(self, config: ProxyCacheConfig, repository: str):
+        self._config = config
+
+        hostname = REGISTRY_URLS.get(
+            config.upstream_registry_hostname,
+            config.upstream_registry_hostname,
+        )
+        url = f"https://{hostname}"
+        if config.insecure:
+            url = f"http://{hostname}"
```

Comment:
Insecure can also mean that the registry is available via `https://`, but its certificates aren't globally trusted/self-signed.

Reply:
As discussed in Slack, we'll handle this as a follow-up. See https://issues.redhat.com/browse/PROJQUAY-3457